src/tree_data.c

```diff
@@ -3365,54 +3380,74 @@ lyd_leafref_link_node_tree(const struct lyd_node *tree)
 
                 if (leaf_schema->type->basetype == LY_TYPE_LEAFREF) {
                     lref = (struct lysc_type_leafref *)leaf_schema->type;
-                    if (lyplg_type_resolve_leafref(lref, elem, &leafref_node->value, tree, &target, &errmsg)) {
+                    if (lyplg_type_resolve_leafref2(lref, elem, &leafref_node->value, tree, &targets, &errmsg)) {
                         /* leafref target not found */
                         free(errmsg);
                     } else {
                         /* leafref target found, link it */
-                        LY_CHECK_RET(lyd_link_leafref_node((struct lyd_node_term *)target, leafref_node));
+                        LY_ARRAY_COUNT_TYPE u;
+
+                        LY_ARRAY_FOR(targets, u) {
+                            if (targets[u]->schema->nodetype & LYD_NODE_TERM) {
+                                ret = lyd_link_leafref_node((struct lyd_node_term *)targets[u], leafref_node);
+                                LY_CHECK_GOTO(ret, cleanup);
+                            }
+                        }
                     }
                 }
             }
             LYD_TREE_DFS_END(sibling, elem);
         }
     }
 
-    return LY_SUCCESS;
+cleanup:
+    LY_ARRAY_FREE(targets);
+    return ret;
 }
 
-LY_ERR
-lyd_unlink_leafref_node(const struct lyd_node_term *node, const struct lyd_node_term *leafref_node)
+/**
+ * @brief Removes links between node and node2.
+ *
+ * @param[in] node Data node to which, the leafref is pointing to.
+ * @param[in] node2 Data node to which, the leafref is pointing to.
+ * @param[in] Whether node2 is target data node (1), or leafref data node (0).
+ * @return LY_SUCCESS on success.
+ * @return LY_ERR value on error.
+ */
+static LY_ERR
+_lyd_unlink_leafref_node(const struct lyd_node_term *node, const struct lyd_node_term *node2, ly_bool target)
```

Comment:
Confusing parameters of the function and since it is used only twice with them switched, why not put the whole functionality into `lyd_unlink_leafref_node()`.

Reply:
The whole point was to avoid code duplication. Do you prefer to have it duplicated within `lyd_unlink_leafref_node`?

Reply:
Yes, I would consider it more readable and ultimately requiring less code.